src/main/scala/controllers/ActionHandler.scala

```diff
@@ -16,32 +17,40 @@ class ActionHandler @Inject()(applications: ApplicationOps, applicationForms: Ap
   import ApplicationData._
   import FieldCheckHelpers._
 
-
   def doSave(id: ApplicationId, sectionNumber: Int, fieldValues: JsObject): Future[Result] = {
-    sectionTypeFor(sectionNumber) match {
-      case VanillaSection =>
-        JsonHelpers.allFieldsEmpty(fieldValues) match {
-          case true => applications.deleteSection(id, sectionNumber).map(_ => redirectToOverview(id))
-          case false => applications.saveSection(id, sectionNumber, fieldValues).map(_ => redirectToOverview(id))
+    gatherSectionDetails(id, sectionNumber).flatMap {
+      case Some(app) =>
+        app.formSection.sectionType match {
+          case SectionTypeForm =>
+            JsonHelpers.allFieldsEmpty(fieldValues) match {
+              case true => applications.deleteSection(id, sectionNumber).map(_ => redirectToOverview(id))
+              case false => applications.saveSection(id, sectionNumber, fieldValues).map(_ => redirectToOverview(id))
+            }
+          case SectionTypeList => Future.successful(redirectToOverview(id))
         }
-      case ItemSection => Future.successful(redirectToOverview(id))
+      case None => Future.successful(NotFound)
     }
   }
 
   def doComplete(id: ApplicationId, sectionNumber: Int, fieldValues: JsObject): Future[Result] = {
-    val answersF: Future[Option[JsObject]] = sectionTypeFor(sectionNumber) match {
-      case VanillaSection => Future.successful(Some(fieldValues))
-      // Instead of using the values that were passed in from the form we'll use the values that
-      // have already been saved against the item list, since these were created by the add-item
-      // form.
-      case ItemSection => applications.getSection(id, sectionNumber).map(_.map(_.answers))
-    }
+    gatherSectionDetails(id, sectionNumber).flatMap {
+      case Some(app) =>
```

Comment:
Well, it is the application, so it's right in that respect. It's just the representation of the application that only contains details of a single section, rather than the entire thing.